optuna/visualization/_regret_bound.py

```diff
@@ -0,0 +1,158 @@
+from __future__ import annotations
+
+from typing import NamedTuple
+
+import tqdm
+
+import optuna
+from optuna._experimental import experimental_func
+from optuna.logging import get_logger
+from optuna.study.study import Study
+from optuna.terminator.erroreval import BaseErrorEvaluator
+from optuna.terminator.erroreval import CrossValidationErrorEvaluator
+from optuna.terminator.improvement.evaluator import BaseImprovementEvaluator
+from optuna.terminator.improvement.evaluator import DEFAULT_MIN_N_TRIALS
+from optuna.terminator.improvement.evaluator import RegretBoundEvaluator
+from optuna.visualization._plotly_imports import _imports
+
+
+if _imports.is_successful():
+    from optuna.visualization._plotly_imports import go
+
+_logger = get_logger(__name__)
+
+
+PADDING_RATIO = 0.05
+OPACITY = 0.25
+
+
+class _RegretBoundInfo(NamedTuple):
+    trial_numbers: list[int]
+    regret_bounds: list[float]
+    errors: list[float] | None
+
+
+@experimental_func("3.2.0")
+def plot_regret_bound(
+    study: Study,
+    plot_error: bool = False,
+    improvement_evaluator: BaseImprovementEvaluator | None = None,
+    error_evaluator: BaseErrorEvaluator | None = None,
+    min_n_trials: int = DEFAULT_MIN_N_TRIALS,
+) -> "go.Figure":
+    _imports.check()
+
+    info = _get_regret_bound_info(study, plot_error, improvement_evaluator, error_evaluator)
+    return _get_regret_bound_plot(info, min_n_trials)
+
+
+def _get_regret_bound_info(
+    study: Study,
+    get_error: bool = False,
+    improvement_evaluator: BaseImprovementEvaluator | None = None,
+    error_evaluator: BaseErrorEvaluator | None = None,
+) -> _RegretBoundInfo:
+    if improvement_evaluator is None:
+        improvement_evaluator = RegretBoundEvaluator()
+    if error_evaluator is None:
+        error_evaluator = CrossValidationErrorEvaluator()
+
+    trial_numbers = []
+    regret_bounds = []
+    errors = []
+
+    for i, trial in enumerate(tqdm.tqdm(study.trials)):
+        trial_numbers.append(trial.number)
+        trials = study.trials[: i + 1]
+        trials = [t for t in trials if t.state == optuna.trial.TrialState.COMPLETE]
+
+        regret_bound = improvement_evaluator.evaluate(
+            trials=trials, study_direction=study.direction
+        )
+        regret_bounds.append(regret_bound)
+
+        if get_error:
+            error = error_evaluator.evaluate(trials=trials, study_direction=study.direction)
+            errors.append(error)
+
+    if len(errors) == 0:
+        return _RegretBoundInfo(
+            trial_numbers=trial_numbers, regret_bounds=regret_bounds, errors=None
+        )
+    else:
+        return _RegretBoundInfo(
+            trial_numbers=trial_numbers, regret_bounds=regret_bounds, errors=errors
+        )
+
+
+def _get_regret_bound_plot(info: _RegretBoundInfo, min_n_trials: int) -> "go.Figure":
```

Comment:
From a readability perspective, I think that there are too many tasks performed in this method, making it difficult to understand its logic at first glance. A single method should focus on a single responsibility (see also: [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single-responsibility_principle)). How about defining a small method with clear and descriptive names for each individual sub-logic, like:
```python
def _get_regret_bound_plot(info: _RegretBoundInfo, min_n_trials: int) -> "go.Figure":
    ...
    
    add_regret_bounds_trace(
        fig, info.trial_numbers[:min_n_trials+1], info.regret_bounds[: min_n_trials + 1],
        n_trials, min_n_trials, plotly_blue_opacity,
    )

    add_regret_bounds_trace(
        fig, info.trial_numbers[min_n_trials:], info.regret_bounds[min_n_trials:],
        n_trials, min_n_trials, plotly_blue,
    )

    add_error_trace(fig, info, plotly_red)
    
    update_y_range(fig, info, min_n_trials)
    
    return fig
```

Reply:
Thank you for your comment. I updated my code by introducing small function extracting some sub-logic. Could you take another look?

Reply:
Thanks for your changes! Now, the function is much more readable for me.